isbg/isbg.py

```diff
@@ -1,909 +1,613 @@
-
-def dehexof(x):
-    res = ""
-    while(len(x)):
-        res = res + chr(16 * hexdigit(x[0]) + hexdigit(x[1]))
-        x = x[2:]
-    return res
-
-# This function makes sure that each lines ends in <CR><LF>
-# SpamAssassin strips out the <CR> normally
-crnlre = re.compile("([^\r])\n", re.DOTALL)
-
-def crnlify(text):
-    # we have to do it twice to work right since the re includes
-    # the char preceding \n
-    return re.sub(crnlre, "\\1\r\n", re.sub(crnlre, "\\1\r\n", text))
-
-def truncate(inp, length):
-    if len(inp) > length:
-        return repr(inp)[:length-3] + '...'
-    else:
-        return inp
-
-def shorten(inp, length):
-    if isinstance(inp, dict):
-        return dict([(k, shorten(v, length)) for k,v in inp.items()])
-    elif isinstance(inp, list) or isinstance(inp, tuple):
-        return [ shorten(x, length) for x in inp]
-    else:
-        return truncate(inp, length)
-
-def imapflags(flaglist):
-    return '(' + ','.join(flaglist) + ')'
-
-class ISBG:
-    exitcodeok = 0          # all went well
-    exitcodenewmsgs = 1     # there were new messages - none of them spam
-    exitcodenewspam = 2     # they were all spam
-    exitcodenewmsgspam = 3  # there were new messages and new spam
-    exitcodeflags = 10      # there were errors in the command line arguments
-    exitcodeimap = 11       # there was an IMAP level error
-    exitcodespamc = 12      # error of communication between spamc and spamd
-    exitcodetty = 20        # error because of non interative terminal
-    exitcodelocked = 30     # there's certainly another isbg running
+    """Class for the ISBG exceptions.
+
+    Attributes:
+        exitcode (int): The exit code for the error.
+        message (str): The human readable error message.
+
+    """
+
+    def __init__(self, exitcode=0, message=""):
+        """Initialize a ISBGError object.
+
+        Args:
+            exitcode (int): Exit code. It must be a
+                :py:const:`~isbg.isbg.__exitcodes__` valid value.
+            message (str): Human readable error message.
+        Raises:
+            ValueError: If the `exitcode` is not in `__exitcodes__`.
+
+        """
+        self.exitcode = exitcode
+        self.message = message
+
+        Exception.__init__(self, message)
+        if exitcode not in __exitcodes__.values():
+            raise ValueError
+
+
+class ISBG(object):
+    """Main ISBG class.
+
+    See also:
+        :py:mod:`isbg.__main__` for command line usage.
+
+    Examples:
+        >>> import isbg
+        >>> sbg = isbg.ISBG()
+        >>> sbg.imapsets.host = "imap.example.org"
+        >>> sbg.imapsets.port = 993
+        >>> sbg.imapsets.user = "example@example.org"
+        >>> sbg.imapsets.passwd = "xxxxxxxx"
+        >>> sbg.imapsets.inbox = "INBOX"
+        >>> sbg.imapsets.spaminbox = "Spam"
+        >>> sbg.imapsets.learnspambox = "Spam"
+        >>> sbg.imapsets.learnhambox = "NOSPAM"
```

Comment:
Just to be consistent, we are using `INBOX.Spam` everywhere in the doc & code as the spambox example.